### PR TITLE
Support for std::span for winrt::array_view and winrt::com_array

### LINKED ISSUE #1289: Allow passing array_view as span
I want to be able to use std::span for portablity and implicitly pass an array_view from WinRT to it.

### PR DESCRIPTION
**Pull request by jonwis**
This fixes #1289 per "option C" by completing what #1331 started.

* Constructors for both array_view and com_array that take a span
* Conversions on both array_view and com_array that produce a span
* Tests for the conversions and constructions
* Additional ctad support for easy use

@JaiganeshKumaran thank you for the initial start - see if this meets your needs, please.

### CODE REVIEW

File: test/test_cpp20/array_span.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+//
+// This test illustrates an array_view<T> (non-const) bound to a std::span<T> on a std::array
+//
+TEST_CASE("array,DataReader,std::span")
+{
+    auto reader = CreateDataReader({ 1, 2, 3 }).get();
+
+    std::array<byte, 3> a{};
+    std::span<byte> sp(a);
+    reader.ReadBytes(sp); // FillArray pattern
```
Comment (sylveon) [inline test/test_cpp20/array_span.cpp:37]: Out of curiosity, are we able to directly pass std::array here now?

File: strings/base_array.h [hunk 2]
```diff
@@ -31,6 +31,20 @@ WINRT_EXPORT namespace winrt
             array_view(value.begin(), static_cast<size_type>(value.size()))
         {}
 
+#ifdef __cpp_lib_span
+        template <typename C, std::size_t extent>
+        array_view(std::span<C, extent> span) noexcept :
+            array_view(span.data(), static_cast<size_type>(span.size())) 
+        {
+            WINRT_ASSERT(span.size() <= UINT_MAX);
```
Comment (JaiganeshKumaran) [inline strings/base_array.h:39]: There are other places (like the constructor of hstring) where currently no check is done. Need to also add this assertion there. Another thing, I would probably recommend using WINRT_VERIFY instead as WINRT_ASSERT is removed in release mode.

Comment (jonwis) [reply strings/base_array.h:39]: Looks like this should have been an `if (too large) { throw invalid arg }` instead. That's what other places do.

Comment (jonwis) [reply strings/base_array.h:39]: But then there's the other places in array_view that are more blase about it:

```c++
        array_view(pointer first, pointer last) noexcept :
            m_data(first),
            m_size(static_cast<size_type>(last - first))
        {}

        array_view(std::initializer_list<value_type> value) noexcept :
            array_view(value.begin(), static_cast<size_type>(value.size()))
        {}
```

@kennykerr what's the overall philosophy here? While it's _super unlikely_ that anybody passes UINT_MAX elements to any WinRT API, I suppose it's not unreasonable...

Comment (jonwis) [reply strings/base_array.h:39]: (And now you've got me worried about truncation errors in string comparisons and other things...)

Comment (kennykerr) [reply strings/base_array.h:39]: C++/WinRT will mostly (debug) assert for things like this. The primary outlier is when an HSTRING initializer is not null terminated in which case it panics. 

I'm not opposed to turning these into exceptions but that would ideally be done consistently throughout the library and consideration be taken about code size impact.

**Comment by kennykerr** [comment]
Hopefully @DefaultRyan can spare a few minutes to look this over. He probably has the best handle on the standard library best practices.

File: strings/base_array.h [hunk 3]
```diff
@@ -223,6 +237,11 @@ WINRT_EXPORT namespace winrt
     template <typename C, size_t N> array_view(std::array<C, N>& value) -> array_view<C>;
     template <typename C, size_t N> array_view(std::array<C, N> const& value) -> array_view<C const>;
 
+#ifdef __cpp_lib_span
+    template <typename C, std::size_t extent> array_view(std::span<C, extent>& value) -> array_view<C>;
```
Comment (kennykerr) [inline strings/base_array.h:241]: cppwinrt uses the undecorated integral types like `size_t` from `stdint.h` so no need for `std...`.

File: strings/base_array.h (hunk 2 above)
Comment (JaiganeshKumaran) [reply strings/base_array.h:39]: The issue of returning span for a local vector could be solved by deleting the rvalue qualified conversion operator but unfortunately, the standard library does not use this mechanism anywhere.

Comment (jonwis) [reply strings/base_array.h:39]: Tracked in #1347 so we can make progress here.

**Review by kennykerr — APPROVED** [review]
I still think the lifetime here is bonkers, but admittedly not any worse than the C++ standard library.

**Comment by dmachaj** [comment]
Sorry I'm a little late to the party.  Do these new methods affect vtable layout at all?  In other words, does span support need an ODR pragma to prevent mismatches?

**Comment by kennykerr** [comment]
> Do these new methods affect vtable layout at all?

`std::span` is not an ABI type.

**Comment by jonwis** [comment]
It's pure conversion methods; no change to the layout or functionality of `array_view` or `com_array` here.

**Pull request merged**